crates/crypto/src/ed25519.rs

```diff
@@ -0,0 +1,136 @@
+use ed25519_dalek::{Signature, VerifyingKey};
+
+use crate::{functions::to_sized, CryptoResult};
+
+const ED25519_PUBKEY_LEN: usize = 32;
+const ED25519_SIGNATURE_LEN: usize = 64;
+
+/// Verify an ED25519 signature with the given hashed message and public
+/// key.
+///
+/// NOTE: This function takes the hash of the message, not the prehash.
+pub fn ed25519_verify(msg_hash: &[u8], sig: &[u8], pk: &[u8]) -> CryptoResult<()> {
+    // Validation
+    let sig = to_sized::<ED25519_SIGNATURE_LEN>(sig)?;
+    let pk = to_sized::<ED25519_PUBKEY_LEN>(pk)?;
+
+    let vk = VerifyingKey::from_bytes(&pk)?;
+
+    vk.verify_strict(msg_hash, &Signature::from(sig))
+        .map_err(Into::into)
+}
+/// Verify a batch of ED25519 signatures with the given hashed message and public
+/// key.
+/// NOTE: This function takes the hash of the messages, not the prehash.
+pub fn ed25519_batch_verify(
+    msgs_hash: &[&[u8]],
+    sigs: &[&[u8]],
+    pks: &[&[u8]],
+) -> CryptoResult<()> {
+    let (sigs, vks): (Vec<Signature>, Vec<VerifyingKey>) = sigs
+        .iter()
+        .zip(pks.iter())
+        .map(|(sig, pk)| -> CryptoResult<(Signature, VerifyingKey)> {
+            let sig = to_sized::<ED25519_SIGNATURE_LEN>(sig)?;
+            let pk = to_sized::<ED25519_PUBKEY_LEN>(pk)?;
+            Ok((Signature::from_bytes(&sig), VerifyingKey::from_bytes(&pk)?))
+        })
+        .collect::<CryptoResult<Vec<(Signature, VerifyingKey)>>>()?
+        .iter()
+        .cloned()
```

Comment:
User `.into_iter()` instead of `.iter().cloned()` to avoid the unnecessary cloning.